internal/controlplane/handlers_user.go

```diff
@@ -259,3 +262,132 @@ func (s *Server) GetUser(ctx context.Context, _ *pb.GetUserRequest) (*pb.GetUser
+			if err != nil {
+				zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
+				return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", sponsorUser.IdentitySubject)
+			}
+		}
+		invitations = append(invitations, &pb.Invitation{
+			Code:           invite.Code,
+			Role:           invite.Role,
+			Email:          invite.Email,
+			Project:        invite.Project.String(),
+			CreatedAt:      timestamppb.New(invite.CreatedAt),
+			ExpiresAt:      timestamppb.New(invite.UpdatedAt.Add(7 * 24 * time.Hour)),
+			Expired:        time.Now().After(invite.UpdatedAt.Add(7 * 24 * time.Hour)),
+			Sponsor:        identity.UserID,
+			SponsorDisplay: identity.Human(),
+		})
+	}
+
+	return &pb.ListInvitationsResponse{
+		Invitations: invitations,
+	}, nil
+}
+
+// ResolveInvitation is a service for resolving an invitation.
+func (s *Server) ResolveInvitation(ctx context.Context, req *pb.ResolveInvitationRequest) (*pb.ResolveInvitationResponse, error) {
+	// Check if the UserManagement feature is enabled
+	if !flags.Bool(ctx, s.featureFlags, flags.UserManagement) {
+		return nil, status.Error(codes.Unimplemented, "feature not enabled")
+	}
+
+	// Extracts the user email from the token
+	tokenString, err := gauth.AuthFromMD(ctx, "bearer")
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no auth token: %v", err)
+	}
+	token, err := s.jwt.ParseAndValidate(tokenString)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to parse bearer token: %v", err)
+	}
+
+	// Check if the invitation code is valid
+	ret, err := s.store.GetInvitationByCode(ctx, req.Code)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "invitation not found")
+		}
+		return nil, status.Errorf(codes.Internal, "failed to get invitation: %s", err)
+	}
+
+	// Check if the invitation matches the user email
+	if token.Email() != ret.Email {
+		return nil, util.UserVisibleError(codes.PermissionDenied, "invitation does not match user")
+	}
+
+	// Check if the invitation is expired
+	if time.Now().After(ret.UpdatedAt.Add(7 * 24 * time.Hour)) {
+		return nil, util.UserVisibleError(codes.PermissionDenied, "invitation expired")
+	}
+
+	isAccepted := false
+	// Accept invitation
+	if req.Accept {
+		// Parse the role
+		authzRole, err := authz.ParseRole(ret.Role)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to parse invitation role: %s", err)
+		}
+		// Add the user to the project
+		if err := s.authzClient.Write(ctx, token.Subject(), authzRole, ret.Project); err != nil {
+			return nil, status.Errorf(codes.Internal, "error writing role assignment: %v", err)
+		}
+		isAccepted = true
+	}
+
+	// Delete the invitation since its resolved
+	ret, err = s.store.DeleteInvitation(ctx, req.Code)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to delete invitation: %s", err)
```

Comment:
What happens in this case -- the user has been added to the OpenFGA database; if they try calling `ResolveInvitation` again, will they create a second tuple?  Will OpenFGA error with "already exists"?

We may want to return a `util.UserVisibleError(codes.Internal, "permission granted, but unable to revoke invitation")` after logging the error internally here.

We may also want to check what the result of attempting to grant the same permission twice is in terms of OpenFGA errors, and account for that specially.